text/0097-log-data-model.md

```diff
@@ -0,0 +1,972 @@
+  * [Trace Context Fields](#trace-context-fields)
+    * [Field: `TraceId`](#field-traceid)
+    * [Field: `SpanId`](#field-spanid)
+    * [Field: `TraceFlags`](#field-traceflags)
+  * [Severity Fields](#severity-fields)
+    * [Field: `SeverityText`](#field-severitytext)
+    * [Field: `SeverityNumber`](#field-severitynumber)
+    * [Displaying Severity](#displaying-severity)
+    * [Comparing Severity](#comparing-severity)
+  * [Field: `ShortName`](#field-shortname)
+  * [Field: `Body`](#field-body)
+  * [Field: `Resource`](#field-resource)
+  * [Field: `Attributes`](#field-attributes)
+* [Example Log Records](#example-log-records)
+* [Open Questions](#open-questions)
+* [Alternate Design](#alternate-design)
+* [Prior Art](#prior-art)
+  * [RFC5424 Syslog](#rfc5424-syslog)
+  * [Fluentd Forward Protocol Model](#fluentd-forward-protocol-model)
+* [Appendix A. Example Mappings.](#appendix-a-example-mappings)
+  * [RFC5424 Syslog](#rfc5424-syslog-1)
+  * [Windows Event Log](#windows-event-log)
+  * [SignalFx Events](#signalfx-events)
+  * [Splunk HEC](#splunk-hec)
+  * [Log4j](#log4j)
+  * [Zap](#zap)
+  * [Apache](#apache)
+  * [CloudTrail Log Event](#cloudtrail-log-event)
+* [Appendix B: `SeverityNumber` example mappings.](#appendix-b-severitynumber-example-mappings)
+* [References](#references)
+
+## Motivation
+
+This is a proposal of a data model and semantic conventions that allow to
+represent logs from various sources: application log files, machine generated
+events, system logs, etc. Existing log formats can be unambiguously mapped to
+this data model. Reverse mapping from this data model is also possible to the
+extent that the target log format has equivalent capabilities.
+
+The purpose of the data model is to have a common understanding of what a log
+record is, what data needs to be recorded, transferred, stored and interpreted
+by a logging system.
+
+## Design Notes
+
+### Requirements
+
+The Data Model was designed to satisfy the following requirements:
+
+- It should be possible to unambiguously map existing log formats to this Data
+  Model. Translating log data from an arbitrary log format to this Data Model
+  and back should ideally result in identical data.
+
+- Mappings of other log formats to this Data Model should be semantically
+  meaningful. The Data Model must preserve the semantics of particular elements
+  of existing log formats.
+
+- Translating log data from an arbitrary log format A to this Data Model and
+  then translating from the Data Model to another log format B ideally must
+  result in a meaningful translation of log data that is no worse than a
+  reasonable direct translation from log format A to log format B.
+
+- It should be possible to efficiently represent the Data Model in concrete
+  implementations that require the data to be stored or transmitted. We
+  primarily care about 2 aspects of efficiency: CPU usage for
+  serialization/deserialization and space requirements in serialized form. This
```

Comment:
does it include schematized /serialization (when schema is stored separately and log entries do not repeat the attribute names)?

I'm asking to understand the assumptions of this document, not recommending features.

Reply:
Yes, partially. For example ProtoBuf-based formats will likely use fields in messages to represent the top-level fields of this data model.  ProtoBuf fields do not repeat attribute names and the schema of ProtoBuf message is not transferred on the wire, it is stored separately.

However, for attributes stored in `Attributes` key/value map or in `Resource`, the names of such attributes are likely going to be transmitted over the wire for every log record.

Reply:
Would it be possible to design the wire protocol in such a way that it is possible to optimize transmission of attribute/key names?  For example, a provision for having a "dictionary" and to store keys as numbers?

Reply:
Possibly, yes. That should be part of protocol discussion and is not in the scope of this OTEP.

Reply:
Answered, resolving.